consensus/beacon/consensus.go

```diff
@@ -341,9 +341,6 @@ func (beacon *Beacon) Finalize(chain consensus.ChainHeaderReader, header *types.
 		amount = amount.Mul(amount, big.NewInt(params.GWei))
 		state.AddBalance(w.Address, amount)
 	}
-	// The block reward is no longer handled here. It's done by the
-	// external consensus engine.
-	header.Root = state.IntermediateRoot(true)
```

Comment:
You removed the IntermediateRoot from `Finalize`. `Finalize` is called from `FinaliseAndAssemble`, where you added a call to IntermediateRoot after the call to `Finalize`. 


`Finalize` is however also called from `state_processor.go:L95`, and there you have not added any corresponding call to intermediate root. How come?

Reply:
In `state_processor.go:L95`, all this function should do is: 
- apply the transactions inside of the block on top on the provided stateDB
- apply the post-process ethereum rules, e.g. pow block rewards, uncle rules, withdrawals, etc
- return the generated receipts and logs
and that all.

`header.Root = state.IntermediateRoot(true)` is just to derive the hash of the mutated stateDB
and assign it to header. In case of ``state_processor.go:L95`, the block is already generated
and retrieved from the network, we don't need to re-generate/assign the new hash to the header.
What's more, the header is just a copied from the block, it's noop.

In block insertion, the real hash comparison is happened in block validation,

```go
	// Validate the state root against the received state root and throw
	// an error if they don't match.
	root, err := statedb.IntermediateRoot(v.config.IsEIP158(header.Number))
	if err != nil {
		return fmt.Errorf("failed to derive state root %v", err)
	}
	if header.Root != root {
		return fmt.Errorf("invalid merkle root (remote: %x local: %x)", header.Root, root)
	}
```

Reply:
In summary, the behavior is identical with old code, but just emphasize 

- for block insertion, `Finalize` is only used to apply extra post-processing rules(e.g. block rewards, withdrawal, etc) but do nothing else.
- for block generation, `FinalizeAndAssemble` is required to both do `Finalize` work and also assemble a block locally,  deriving the state hash from the stateDB and wrap it in a block.